application/btsieve/src/bitcoin/queries/transaction.rs

```diff
@@ -22,31 +21,55 @@ impl QueryType for TransactionQuery {
     }
 }
 
-impl ShouldExpand for TransactionQuery {
-    fn should_expand(query_params: &QueryParams) -> bool {
-        query_params.expand_results
-    }
+#[derive(Deserialize, Derivative, Debug, PartialEq)]
+#[derivative(Default)]
+#[serde(rename_all = "snake_case")]
+pub enum ReturnAs {
+    #[derivative(Default)]
+    TransactionId,
+    Transaction,
 }
 
-impl ExpandResult for TransactionQuery {
+impl ToHttpPayload<TransactionQuery, ReturnAs> for QueryResult {
     type Client = BitcoinCoreClient;
-    type Item = Transaction;
-
-    fn expand_result(
-        result: &QueryResult,
-        client: Arc<BitcoinCoreClient>,
-    ) -> Result<Vec<Transaction>, Error> {
-        let mut expanded_result: Vec<Transaction> = Vec::new();
-        for tx_id in result.clone().0 {
-            let tx_id = TransactionId::from_hex(tx_id.as_str()).map_err(|_| Error::InvalidHex)?;
+    type Item = PayloadKind;
+
+    fn to_http_payload(
+        &self,
+        return_as: &ReturnAs,
+        client: &BitcoinCoreClient,
+    ) -> Result<Vec<Self::Item>, Error> {
+        // Close over some local variables for easier usage of the method
+        let to_payload = |id: TransactionId| convert_to_payload(client, return_as, id);
+
+        self.0
+            .iter()
+            .filter_map(to_sha256d_hash)
+            .map(to_payload)
+            // .collect for Vec<Result<PayloadKind, Error>> transforms it into
+            // Result<Vec<PayloadKind>, Error> returning the first Error or the whole
+            // collection.
+            // We want this because the the Error means something is wrong with the connection to
+            // our bitcoin node and skipping is unlikely to help since the next call will fail as
+            // well. That is why we simply fail the whole function.
+            .collect()
+    }
+}
 
-            let transaction = client
-                .get_raw_transaction_verbose(&tx_id)
-                .map_err(Error::BitcoinRpcConnection)?
-                .map_err(Error::BitcoinRpcResponse)?;
-            expanded_result.push(transaction.into());
-        }
-        Ok(expanded_result)
+fn convert_to_payload(
+    client: &BitcoinCoreClient,
+    return_as: &ReturnAs,
+    id: TransactionId,
+) -> Result<PayloadKind, Error> {
+    match return_as {
+        ReturnAs::TransactionId => Ok(PayloadKind::Id { id }),
+        ReturnAs::Transaction => match client.get_raw_transaction_verbose(&id) {
+            Ok(Ok(transaction)) => Ok(PayloadKind::Transaction {
+                transaction: transaction.into(),
+            }),
+            Ok(Err(e)) => Err(Error::BitcoinRpcResponse(e)),
```

Comment:
Do you mean instead of erroring out? I don't think so since whoever calls this endpoint has expectations of what is returned because they set the query parameter. Returning something different will break them at runtime ...